python/ray/tune/utils/file_transfer.py

```diff
@@ -0,0 +1,236 @@
+    force_all: bool = False,
+    chunk_size: int = DEFAULT_CHUNK_SIZE,
+    _return_all_remotes: bool = False,
+) -> Union[ray.ObjectRef, Tuple[ray.ObjectRef, ray.ActorID, ray.ObjectRef]]:
+    """Synchronize directory on source node to directory on target node.
+
+    Per default, this function will collect information about already existing
+    files in the target directory. Only files that differ in either mtime or
+    filesize will be transferred, unless ``force_all=True``.
+
+    Args:
+        source_ip: IP of source node.
+        source_path: Path to file or directory on source node.
+        target_ip: IP of target node.
+        target_path: Path to file or directory on target node.
+        force_all: If True, all files will be transferred (not just differing files).
+        chunk_size: Chunk size for data transfer.
+
+    Returns:
+        Ray future for scheduled unpacking task.
+
+    """
+    pack_actor_on_source_node = _PackActor.options(
+        num_cpus=0, resources={f"node:{source_ip}": 0.01}
+    )
+    unpack_on_target_node = _unpack_from_actor.options(
+        num_cpus=0, resources={f"node:{target_ip}": 0.01}
+    )
+
+    if force_all:
+        files_stats = None
+    else:
+        files_stats = _remote_get_recursive_files_and_stats.options(
+            num_cpus=0, resources={f"node:{target_ip}": 0.01}
+        ).remote(target_path)
+
+    pack_actor = pack_actor_on_source_node.remote(
+        source_path, files_stats, chunk_size=chunk_size
+    )
+    unpack_future = unpack_on_target_node.remote(pack_actor, target_path)
+
+    if _return_all_remotes:
+        return unpack_future, pack_actor, files_stats
+
+    return unpack_future
+
+
+def delete_on_node(node_ip: str, path: str) -> ray.ObjectRef:
+    """Delete path on node.
+
+    Args:
+        node_ip: IP of node to delete path on.
+        path: Path to delete on remote node.
+
+    Returns:
+        Ray future for scheduled delete task.
+    """
+    delete_task = _delete_path.options(num_cpus=0, resources={f"node:{node_ip}": 0.01})
+    return delete_task.remote(path)
+
+
+def _get_recursive_files_and_stats(path: str) -> Dict[str, Tuple[float, int]]:
+    """Return dict of files mapping to stats in ``path``.
+
+    This function scans a directory ``path`` recursively and returns a dict
+    mapping each contained file to a tuple of (mtime, filesize).
+
+    mtime and filesize are returned from ``os.lstat`` and are usually a
+    floating point number (timestamp) and an int (filesize in bytes).
+    """
+    files_stats = {}
+    for root, dirs, files in os.walk(path, topdown=False):
+        rel_root = os.path.relpath(root, path)
+        for file in files:
+            key = os.path.join(rel_root, file)
+            stat = os.lstat(os.path.join(path, key))
+            files_stats[key] = stat.st_mtime, stat.st_size
+
+    return files_stats
+
+
+# Only export once
+_remote_get_recursive_files_and_stats = ray.remote(_get_recursive_files_and_stats)
```

Comment:
I saw this was from the existing code, but for my learning do you know what the purpose of doing this is?

Reply:
The reason to do this is to only export the remote function once. If you do this in the public API, you'll export the same method multiple times. Think of this as "caching" the remote wrapper

Reply:
(just fyi if the same method gets exported too many times Ray will raise a warning)

Reply:
Ah I see! Why not use the decorator pattern?

Reply:
In a previous iteration we also directly called the non-remote version, but this was removed - I'll update the PR

Reply:
Actually I'd like to keep both methods, as I might be accessing the non-remote method in a follow up PR. Also since this is private scope it should be ok to have both here